### PR TITLE
[Patterns]: Pattern Directory Explorer

### LINKED ISSUE #44611: Patterns: Connect "Explore" to pattern directory
Issue #41379 (underway in #44028) brings a new flyout to the patterns tab of the inserter:

185140173-deb7255e-53af-4447-9a3c-6644b33806f3

In #44585 this is further enhanced by zooming out the main canvas when the tab is selected, and #44501 curates the pattern categories and which are included. These incoming changes provide good context for reconsidering the "Explore" modal, which currently shows the same patterns that are shown in the main inserter, providing limited utility beyond the more convenient overview.

| Inserter | Explore modal |
| --- | --- |
| inserter | explore modal |

This provides an opportunity to recast the modal to instead become an access point to [the WordPress Pattern Directory](https://wordpress.org/patterns/). It could look like this:

Pattern directory

* Show search and categories in the left sidebar as per usual (see also #44501)
* Shows "Pattern Directory" as title
* Shows filter button in the top corner to toggle between "Newest" and "Popular"

### PR DESCRIPTION
**Pull request by ntsekouras**
## What?
Resolves: https://github.com/WordPress/gutenberg/issues/44611

Quick and dirty POC for adding a Pattern Directory Explorer. This will need much polishing but here it is to start addressing issues that come up and discuss. Check the below notes sections for some initial issues and know issues besides the obvious that takes some time to show results etc..

## Pattern Directory Notes
1. It seems there is no pagination supported in Pattern Directory. Ideally this needs to be handled there and integrate the pagination here. 
2. Another really important issue is that if we make any request to the PD, the eligible patterns to be inserted are subject to the `allowedBlocks` in the editor or a block. What this means is that we could fetch 10 patterns for example where none of them could be inserted.. We must find a way to make this work.. We could explore how we could integrate some of these data to the PD itself(something like `blockTypes` in combination with the allowedBlockTypes passed in the request). Not sure yet..
3. We should match the PD order/orderBy params to show the same results.
4. Are there any other PD APIs available, like `favorite`?
5. How is localization is handled in PD? Both in pattern categories and pattern titles etc..
6. I have noticed that in some patterns images cannot be loaded(403 errors).

## Block editor Notes
1. This might be impactful in terms of performance negatively. We need to measure and find a good implementation, as parsing patterns and then rendering them in Block Previews is expensive. We might consider custom caching, use of stores or even something related to [this PR](https://github.com/WordPress/pattern-directory/pull/528) that uses screenshots of the patterns and is a very interesting idea.
2. In my initial commit I also don't check the fetched patterns if they contain a block that is not allowed. We do that for the registered patterns and is also kind of expensive as we recursively check each pattern's block. 
3. Check to abort ongoing stale requests when needed..

## Testing Instructions
1. Open the `patterns` tab in the main inserter and then click the `Pattern Directory` button.
7. Search some patterns or select categories and insert them..

## Screenshots or screencast 

https://user-images.githubusercontent.com/16275880/197474865-033370e0-315f-48cb-adbd-3a548e77301d.mov

### CODE REVIEW

**Comment by jasmussen** [comment]
Nice! Thanks for working on this. Here's a gif showing the patterns tab and buttons:

explorer

First off, I can't get the connection to work, not sure why that is. But I'm having some local env issues, that might be it.

The other thing, there should not be two buttons, there should be only a single button, which we can call "Explore patterns" at the bottom. This button opens the pattern directory in the modal. We do not need to show any of the locally bundled patterns at all in this modal, as they are all visible from the inserter tab.

**Comment by ntsekouras** [comment]
>First off, I can't get the connection to work, not sure why that is. But I'm having some local env issues, that might be it.

@jasmussen you'll have to give it some time because it makes requests to fetch the pattern categories and then the patterns (there is no pagination so it fetches `100`) and then parse them and render them (not 100 but 10, but still..) 😓 . Eventually there will be loaders to indicated the above.. Please let me know if there are no results after waiting..

>The other thing, there should not be two buttons, 

Yes, I know. I just have it there for now to compare UIs easier. It will be removed in the end.

**Comment by jasmussen** [comment]
> Please let me know if there are no results after waiting..

I let it sit for a few minutes, and kept getting the "No results", but with an endless stream of console errors:

Screenshot 2022-10-24 at 12 44 41

**Comment by ntsekouras** [comment]
That's really weird @jasmussen... There are other errors for me that do not affect all the patterns and the results(403 for some images), but not like the ones you have.. 🤔 .

Also for the record there are no patterns in PD with [test](https://wordpress.org/patterns/search/test/) 😄 . Try something like `heading`.

File: lib/compat/wordpress-6.2/class-gutenberg-rest-pattern-directory-controller-6-2.php [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+			return $raw_patterns;
+		}
+
+		$response = array();
+
+		if ( $raw_patterns ) {
+			foreach ( $raw_patterns as $pattern ) {
+				$response[] = $this->prepare_response_for_collection(
+					$this->prepare_pattern_category_for_response( $pattern, $request )
```
Comment (anton-vlasenko) [inline lib/compat/wordpress-6.2/class-gutenberg-rest-pattern-directory-controller-6-2.php]: `$request` variable is not defined.
It causes this fatal error:
```
PHP Fatal error:  Uncaught Error: Call to a member function has_param() on null in /src/wp-includes/rest-api/endpoints/class-wp-rest-controller.php:591
```

**Review by anton-vlasenko — CHANGES_REQUESTED** [review]

File: lib/compat/wordpress-6.2/class-gutenberg-rest-pattern-directory-controller-6-2.php [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+					'callback'            => array( $this, 'get_pattern_categories' ),
+					'permission_callback' => array( $this, 'get_items_permissions_check' ),
+				)
+			),
+		);
+
+		parent::register_routes();
+	}
+
+	public function get_pattern_categories() {
```
Comment (ryelle) [inline lib/compat/wordpress-6.2/class-gutenberg-rest-pattern-directory-controller-6-2.php]: This fixes the error that prevents categories from loading

```suggestion
	public function get_pattern_categories( $request ) {
```

**Comment by ryelle** [comment]
Yay for browsing the pattern directory from the editor :tada:

I poked around a bit (once I fixed the `$request` issue mentioned). It feels very slow and a bit clunky - hopefully if that's an API issue, we can sort it out with some of the suggestions you've made.

Regarding your pattern directory notes… the current API is a proxy to the REST endpoint for the patterns CPT, so any query you can do on posts/pages/etc we can probably make happen in this API. For example, [patterns by the wordpressdotorg user](https://wordpress.org/patterns/wp-json/wp/v2/wporg-pattern?author=5911429). If you need anything specific, [an issue in `pattern-directory`](https://github.com/WordPress/pattern-directory/) would be best.

I've added an issue about the pagination. https://github.com/WordPress/pattern-directory/issues/529

> Are there any other PD APIs available, like favorite?

Not currently, but it can be added. You'll need to ask for the user's wporg username, like on the Plugins screen (`/wp-admin/plugin-install.php?tab=favorites`), and we can respond back with that user's favorite patterns. Added an issue https://github.com/WordPress/pattern-directory/issues/531

> How is localization is handled in PD? Both in pattern categories and pattern titles etc..

The API layer in core should pass along the site's locale when fetching patterns, so you should see translated pattern titles and content. The category slugs stay the same. It's possible to get translated category names by using the underlying pattern API directly, but I don't think it's possible with the core endpoints yet. Added an issue https://github.com/WordPress/pattern-directory/issues/530

Here's browsing the directory from a Spanish site:
Screen Shot 2022-10-24 at 5 51 05 PM

> We should match the PD order/orderBy params to show the same results.

On WordPress.org we order by date or by number of favorites, is that what you mean here?

> I have noticed that in some patterns images cannot be loaded(403 errors).

I wonder if you're hitting this issue https://github.com/WordPress/pattern-directory/issues/488 — are the 403s from rawpixel?

**Comment by ironprogrammer** [comment]
🎉 I'm confirming that @ryelle's fix to `class-gutenberg-rest-pattern-directory-controller-6-2.php:27` resolves the fatal that @anton-vlasenko identified [here](https://github.com/WordPress/gutenberg/pull/45237#discussion_r1003792681), and the patterns load in "Explore all patterns" (not testing the other "Pattern Directory" button, per @jasmussen's [comment here](https://github.com/WordPress/gutenberg/pull/45237#issuecomment-1288776400)).

File: lib/compat/wordpress-6.2/class-gutenberg-rest-pattern-directory-controller-6-2.php (hunk 1 above)
Comment (ntsekouras) [reply lib/compat/wordpress-6.2/class-gutenberg-rest-pattern-directory-controller-6-2.php]: Thank you Anton!

**Comment by ntsekouras** [comment]
Thanks for the help @anton-vlasenko , @ryelle , @ironprogrammer !

>On WordPress.org we order by date or by number of favorites, is that what you mean here?

Yes, because I noticed that the results are not the same between this PR and the PD.

>I have noticed that in some patterns images cannot be loaded(403 errors).

>   >I wonder if you're hitting this issue https://github.com/WordPress/pattern-directory/issues/488 — are the 403s from rawpixel?

Yes, it seems like this would be it.

In general the pagination issue is the most important one for me to be implemented in Pattern Directory. Thanks for creating the issue.

**Comment by ntsekouras** [comment]
Another really important issue is that if we make any request to the PD, the eligible patterns to be inserted are subject to the `allowedBlocks` in the editor or a block. What this means is that we could fetch 10 patterns for example where none of them could be inserted.. We must find a way to make this work.. We could explore how we could integrate some of these data to the PD itself(something like `blockTypes` in combination with the allowedBlockTypes passed in the request). Not sure yet..

File: lib/compat/wordpress-6.2/class-gutenberg-rest-pattern-directory-controller-6-2.php (hunk 1 above)
Comment (anton-vlasenko) [reply lib/compat/wordpress-6.2/class-gutenberg-rest-pattern-directory-controller-6-2.php]: You are welcome, Nik.

**Comment by ryelle** [comment]
> In general the pagination issue is the most important one for me to be implemented in Pattern Directory. Thanks for creating the issue.

Pagination should now work as it does in other REST API endpoints, though `per_page` still defaults to 100 (so the existing API usage is consistent). I also removed the forced `order`/`orderBy`, so the API returns sorted by newest first (like the pattern dir homepage), or you can use `orderby=favorite_count` to mirror the "Popular" sorting option.

**Comment by ryelle** [comment]
I realized the core endpoint (`/wp-json/wp/v2/pattern-directory/patterns/`) didn't pass those pagination & order parameters through to api.w.org, so I've opened a PR #45293 to add support for pagination to the local site's pattern API.

**Comment by ntsekouras** [comment]
**PR Update**

The missing pieces/questions are:
1. We're about to make the full connection with Pattern Directory [with this PR](https://github.com/WordPress/pattern-directory/pull/540). This is needed for the REST API to fetch only patterns based on the allowed blocks of the insertion point.  
2. For now if you select a category, it fetches `10` patterns. Do we want a different number? Should we add a `load more` button somewhere? If yes it would need design and we could probably do that in a follow up..

**Comment by jameskoster** [comment]
I'd say we'll definitely need a way to load more patterns. A dedicated button is a simple solution, it could be interesting to explore an option the auto-loads the next _x_ patterns as you scroll. Would that be feasible?

**Comment by ntsekouras** [comment]
>it could be interesting to explore an option the auto-loads the next x patterns as you scroll. Would that be feasible?

I think this has many a11y implications and should be avoided. For example it cannot be triggered from screen readers I think.

Another solution would be to have a pagination with `next page` and `previous page` instead of a `load more` button, to reduce the patterns shown . I'm not mentioning `numbers` as well, because I'll have to check what is available from pattern directory API and either way we could do it in a follow up.

**Comment by jameskoster** [comment]
We can probably work around that by having a visually hidden button? But it may not be preferable anyway. A simple button definitely seems like a good place to start. We can explore more elaborate treatments later.

**Comment by alexstine** [comment]
Accessibility feedback.
1. Pagination is a must in this type of situation.
2. Load more buttons should be avoided because the focus regressions are a super pain to prevent.
3. I imagine something like this. Previous Patterns Page, 1-5, the top page, 1 2 3 4 5 30, and then Next Patterns Page.
4. Pagination in React always means some type of focus management or announcement letting users know a new page is loaded.
5. No hidden buttons. It must be worked in to the design.
6. No auto scroll. I cannot speak much to it now as I do not have vision anymore, but those things used to really annoy me when I did. If I tried to scroll down so I could get a better view of the window and it did not work, sometimes it would take me a while to realize I was scrolling a container and not a page. This is a big turn off for people using magnification software.

Keep the same UI for everyone, super easy to make it accessible. I would consider this a major blocker before it goes in to WP Core.

Thanks.

**Comment by ntsekouras** [comment]
Awesome feedback @alexstine, thanks! I wanted to add for the design input one more thing. For start at least, the pagination should be just two buttons with `previous page` and `next page`. This is because the current approach is to be tested, but we might end up with client validation and pagination, so there would be no way to know how many pages exist etc..

**Comment by alexstine** [comment]
@ntsekouras For argument sake, you would still need to know how many items are returned in each paginated call unless you plan on showing next/previous buttons that do not work. For example, if you have 10 items per page with a total of 25 items, that means user should only be able to click next twice. You start on first 10, go to second 10, and end with the last 5. Does the next button not disappear or disable?

**Comment by ntsekouras** [comment]
>@ntsekouras For argument sake, you would still need to know how many items are returned in each paginated call unless you plan on showing next/previous buttons that do not work.

I'm not really sure depending on the final approach. If we do the pagination client side due to the implications of `allowedBlocks per root client id` there would be no way to know the pages from the REST API. In my mind worst case scenario would be click the `next` button and eventually show/announce no more results.  Also I had in mind to disable the buttons - for example when you're in page one the `previous` would be disabled. What's preferable a11y-wise, to disable or hide the buttons?

**Comment by alexstine** [comment]
@ntsekouras Probably just disable would be okay. You can pass the `disabled` prop to the `Button` component and it should work fine.

**Comment by apeatling** [comment]
@ntsekouras Checking in on the status of this one. Is it still something you're working on? I came in to test and it's working well so far. Happy to chip in or help test anything new. 👍

**Comment by ntsekouras** [comment]
>@ntsekouras Checking in on the status of this one. Is it still something you're working on? I came in to test and it's working well so far. Happy to chip in or help test anything new. 👍

@apeatling since [the PD part is merged](https://github.com/WordPress/pattern-directory/pull/540), I updated this PR. Reviews and testing would be helpful, yes :) .

Noting that I noticed [this behavior](https://github.com/WordPress/pattern-directory/pull/540#issuecomment-1375377118):
>I'm testing the GB PR and it seems can fetch patterns with more blocks, if we also provide the `search` param, ex: 
```
https://api.wordpress.org/patterns/1.0/?search=image&allowed_blocks[]=core/heading&allowed_blocks[]=core/gallery
```

**Comment by apeatling** [comment]
Did a quick first pass on the UI, I'll have some more feedback. Two things I noticed so far:

1. It would be great to automatically select the "Featured" category rather than having nothing selected when the modal first pops up. 
2. When I insert a pattern the notice in the bottom left covers the exact spot where the "Explore all patterns" button is, meaning I'm hindered when trying to go back and add another pattern. It would be good to move these notices above then button, or remove them more quickly if the insertion sidebar is opened.


Screenshot 2023-01-11 at 4 04 19 PM

**Review by apeatling — COMMENTED** [review]
Did a quick first pass on the UI, I'll have some more feedback. Two things I noticed so far:

1. It would be great to automatically select the "Featured" category rather than having nothing selected when the modal first pops up. 
2. When I insert a pattern the notice in the bottom left covers the exact spot where the "Explore all patterns" button is, meaning I'm hindered when trying to go back and add another pattern. It would be good to move these notices above the button, or remove them more quickly if the insertion sidebar is opened.


Screenshot 2023-01-11 at 4 04 19 PM

**Comment by jasmussen** [comment]
> or remove them more quickly if the insertion sidebar is opened.

A sidebar, but I've noticed these linger for a very very long time. I'd love to reduce the time they stay on screen.

**Comment by apeatling** [comment]
> A sidebar, but I've noticed these linger for a very very long time. I'd love to reduce the time they stay on screen.

Agree -- the messages are usually short so no need for them to linger.

**Comment by apeatling** [comment]
@jasmussen I've opened a PR to half the snackbar display time in #47199.

**Pull request closed without merging**